pallets/moonbeam-lazy-migrations/src/lib.rs

```diff
@@ -65,6 +86,213 @@ pub mod pallet {
 		AddressesLengthCannotBeZero,
 		/// The contract is not corrupted (Still exist or properly suicided)
 		ContractNotCorrupted,
+		/// The key lengths exceeds the maximum allowed
+		KeyTooLong,
+	}
+
+	pub(crate) const MAX_ITEM_PROOF_SIZE: u64 = 30 * 1024; // 30 KB
+	pub(crate) const PROOF_SIZE_BUFFER: u64 = 100 * 1024; // 100 KB
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_n: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			let proof_size_before: u64 = get_proof_size().unwrap_or(0);
+			let res = Pallet::<T>::handle_migration(remaining_weight);
+			let proof_size_after: u64 = get_proof_size().unwrap_or(0);
+			let proof_size_diff = proof_size_after.saturating_sub(proof_size_before);
+
+			Weight::from_parts(0, proof_size_diff)
+				.saturating_add(T::DbWeight::get().reads_writes(res.reads, res.writes))
+		}
+	}
+
+	#[derive(Default, Clone, PartialEq, Eq, Encode, Decode, Debug)]
+	pub(crate) struct ReadWriteOps {
+		pub reads: u64,
+		pub writes: u64,
+	}
+
+	impl ReadWriteOps {
+		pub fn new() -> Self {
+			Self {
+				reads: 0,
+				writes: 0,
+			}
+		}
+
+		pub fn add_one_read(&mut self) {
+			self.reads += 1;
+		}
+
+		pub fn add_one_write(&mut self) {
+			self.writes += 1;
+		}
+
+		pub fn add_reads(&mut self, reads: u64) {
+			self.reads += reads;
+		}
+
+		pub fn add_writes(&mut self, writes: u64) {
+			self.writes += writes;
+		}
+	}
+
+	#[derive(Clone)]
+	struct StateMigrationResult {
+		last_key: Option<StorageKey>,
+		error: Option<&'static str>,
+		reads: u64,
+		writes: u64,
+	}
+
+	enum NextKeyResult {
```

Comment:
nit: in Rust I would usually expect that a Result is something containing Ok() and Error(), it was weird for me reading three variants, but just a comment

Reply:
At first I modeled this as `Result<Option<Key>, Err>` but it required me to write more nested code. So I changed it to be this way, which will result it better code.

Btw I found an example from substrate where this kind of enum exists: https://paritytech.github.io/polkadot-sdk/master/sc_consensus/block_import/enum.ImportResult.html

Reply:
Great! thanks for the insight